src/net_processing.cpp

```diff
@@ -3236,9 +3314,46 @@ bool PeerLogicValidation::SendRejectsAndCheckIfBanned(CNode* pnode, bool enable_
     return false;
 }
 
+bool IsPendingBlockValidation(CNode* pfrom, CPeerState* peerstate) EXCLUSIVE_LOCKS_REQUIRED(cs_peerstate)
+{
+    if (peerstate->pending_block_processing.valid()) {
+        if (peerstate->pending_block_processing.wait_for(std::chrono::duration<int>::zero()) == std::future_status::ready) {
+            bool fNewBlock = peerstate->pending_block_processing.get();
+            if (fNewBlock) {
+                pfrom->nLastBlockTime = GetTime();
+            } else {
+                LOCK(cs_main);
+                mapBlockSource.erase(peerstate->pending_block_hash);
+            }
+            peerstate->pending_block_processing = std::future<bool>();
+            peerstate->pending_block_hash = uint256();
+
+            // We need to wait for BlockChecked to fire, so we don't re-enable
+            // this peer until the current pending validationinterface callbacks
+            // drain. This ensures bans and reject messages happen in order, as
+            // expected by some of our functional tests.
+            peerstate->pending_event_wait = true;
+            NodeId node_id = pfrom->GetId();
+            CallFunctionInValidationInterfaceQueue([node_id] {
+                LOCK(cs_peerstate);
+                CPeerState* peerstate = PeerState(node_id);
```

Comment:
This `peerstate` shadows the `peerstate` in the outer context. Consider renaming :-)

Reply:
It's not captured so does is it really shadows?